homeassistant/helpers/device.py

```diff
@@ -0,0 +1,86 @@
+"""Provides useful helpers for handling devices."""
+
+from homeassistant.core import HomeAssistant, callback
+
+from . import device_registry as dr, entity_registry as er
+
+
+@callback
+def async_entity_id_to_device_id(
+    hass: HomeAssistant,
+    entity_id_or_uuid: str,
+) -> str | None:
+    """Resolve the device id to the entity id or entity uuid."""
+
+    ent_reg = er.async_get(hass)
+
+    entity_id = er.async_validate_entity_id(ent_reg, entity_id_or_uuid)
+    if (entity := ent_reg.async_get(entity_id)) is None:
+        return None
+
+    return entity.device_id
+
+
+@callback
+def async_device_info_to_link_entity(
```

Comment:
This is the only method called; all others added; I feel like adding noise to my taste and makes code readability harder. As a side effect, this thus adds generic helpers we otherwise aren't using.

I also think the method is incorrect named:

> device info to link entity

Sounds not fully right in terms of natural language.

> link entity to device

does.

With that in mind, it also makes me wonder, is this a device helper? Or a helper for entities to link them up to devices?

Reply:
This helper returns DeviceInfo with just the information to link the device to an entity and the configuration entry is classified as "link". This helper fetches information from an entity's device.

`async_device_info_to_link_device_id` does the same thing by looking for information from a device ID.

Would `async_device_info_link_from_device_id` and `async_device_info_link_from_entity` be better?

Reply:
The `async_device_info_to_link_from_device_id` helper is not yet used, but there are use cases in the PRs that will come to use them.

For example, they will be used in #117753 and #118008.

Reply:
`async_device_info_to_link_from_device_id` Then it should be removed from this PR.

PS should be reduced to their bare minimum, and show their use cases.
From the point of this specific PR, it introduced unused code.

Reply:
Removed